toolchain/check/check.cpp

```diff
@@ -56,16 +85,214 @@ auto CheckParseTree(SharedValueStores& value_stores,
+        }
+        continue;
+      }
+
+      if (package->api_or_impl == Parse::Tree::ApiOrImpl::Impl) {
+        continue;
+      }
 
-  return sem_ir;
+      auto [entry, success] = api_map.insert({package_key, &unit_info});
+      if (!success) {
+        // TODO: Cross-reference the source, deal with library, etc.
+        CARBON_DIAGNOSTIC(DuplicateLibraryApi, Error,
+                          "Library's API declared in more than one file.");
+        unit_info.emitter.Emit(package->package.node, DuplicateLibraryApi);
+      }
+    }
+  }
+
+  // TODO: Detect self import.
+  // Mark down imports for all files.
+  llvm::SmallVector<UnitInfo*> ready_to_check;
+  ready_to_check.reserve(units.size());
+  for (auto& unit_info : unit_infos) {
+    const auto& package = unit_info.unit->parse_tree->package();
+    if (package && package->api_or_impl == Parse::Tree::ApiOrImpl::Impl) {
+      // An `impl` has an implicit import of its `api`.
+      TrackImport(api_map, nullptr, unit_info, package->package);
+    }
+
+    llvm::DenseMap<ImportKey, Parse::Node> explicit_import_map;
+    for (const auto& import : unit_info.unit->parse_tree->imports()) {
+      TrackImport(api_map, &explicit_import_map, unit_info, import);
+    }
+
+    // If there were no imports, mark the file as ready to check for below.
+    if (unit_info.imports_remaining == 0) {
+      ready_to_check.push_back(&unit_info);
+    }
+  }
+
+  // Check everything with no dependencies. Earlier entries with dependencies
+  // will be checked as soon as all their dependencies have been checked.
+  for (int check_index = 0;
+       check_index < static_cast<int>(ready_to_check.size()); ++check_index) {
+    auto* unit_info = ready_to_check[check_index];
+    CheckParseTree(builtin_ir, *unit_info, vlog_stream);
+    for (auto* incoming_import : unit_info->incoming_imports) {
+      --incoming_import->imports_remaining;
+      if (incoming_import->imports_remaining == 0) {
+        ready_to_check.push_back(incoming_import);
+      }
+    }
+  }
+
+  // If there are still units with remaining imports, it means there's a
+  // dependency loop.
+  if (ready_to_check.size() < unit_infos.size()) {
+    // Go through units and mask out unevaluated imports. This breaks everything
+    // associated with a loop equivalently, whether it's part of it or depending
+    // on a part of it.
+    // TODO: Better identify cycles, maybe try to untangle them.
+    for (auto& unit_info : unit_infos) {
+      if (unit_info.imports_remaining > 0) {
+        for (auto* import_it = unit_info.imports.begin();
+             import_it != unit_info.imports.end();) {
+          auto* import_unit = import_it->second->unit;
+          if (*import_unit->sem_ir) {
+            ++import_it;
+          } else {
+            CARBON_DIAGNOSTIC(
+                ImportCycleDetected, Error,
+                "Import is in cycle. Cycle must be fixed to import.");
```

Comment:
The imports we identify here aren't necessarily in a cycle. Eg, if we have A imports B imports C imports D imports C, then only C and D are in a cycle, but this will identify A's import of B as well, which is not in a cycle, but can't be imported because of a cycle.

Maybe something like "Library cannot be imported due to a cycle." or "Import cannot be performed due to a cycle." would be more accurate than "Import is in a cycle."?